src/poetry/puzzle/provider.py

```diff
@@ -460,18 +460,29 @@ def complete_package(
             dependency = dependency_package.dependency
             requires = package.requires
 
-        optional_dependencies = []
+        found_extras = set()
+        optional_dependencies = set()
         _dependencies = []
 
-        # If some extras/features were required, we need to
-        # add a special dependency representing the base package
-        # to the current package
         if dependency.extras:
-            for extra in dependency.extras:
-                if extra not in package.extras:
+            # Find all the optional dependencies that are wanted - taking care to allow
+            # for self-referential extras.
+            stack = list(dependency.extras)
+            while stack:
+                extra = stack.pop()
+                if extra in found_extras:
                     continue
+                found_extras.add(extra)
 
-                optional_dependencies += [d.name for d in package.extras[extra]]
+                extra_dependencies = package.extras.get(extra, [])
+                for extra_dependency in extra_dependencies:
+                    if extra_dependency.name == dependency.name:
+                        stack += list(extra_dependency.features)
```

Comment:
```suggestion
                        stack += list(extra_dependency.extras)
```

Can we stick with `extras` here? (This should be the same.)

Reply:
Aren't features "activated extras" as opposed to "available extras"?

Reply:
https://github.com/python-poetry/poetry-core/blob/afaa1285f160f0bad2ee9c3aa6133498ce3245af/src/poetry/core/packages/dependency.py#L197-L200

looks like a distinction without difference